os/kernel/sched: update cpuload measurement for smp

ritesh55555 opened the pull request:
This commit calculates cpu load for each core separately whereas previously it was only calculating for one core. This gives the cpu load percentage of each thread for their respective cores.

It also outputs only threads with more than 0 cpu loads

```
	  The previous output when when we tash command as "cpuload" was : 
          TASH>> cpuload Started CPU monitor with interval 5.
	  PID | Pri |    2s |
	  --------------------------------------------------
	    0 |   0 |  99.1 | Idle Task
  	    1 | 224 |   0.0 | hpwork
 	    2 | 113 |   0.0 | lpwork
	    3 | 120 |   0.0 | EHCI Monitor
	    4 | 110 |   0.0 | LWIP_TCP/IP
  	    5 | 250 |   0.0 | binary_manager
   	    7 | 220 |   0.0 | /dev/mtdblock2
  	    8 | 221 |   0.0 | msg_receiver
	    9 | 221 |   0.0 | multi_recv_nonblock
	   10 | 221 |   0.0 | multi_recv_block1
	   11 | 221 |   0.0 | multi_recv_block2
	   12 | 180 |   0.0 | /dev/mtdblock4
	   13 | 100 |   0.0 | uwork
	   14 | 125 |   0.9 | tash
	   15 | 100 |   0.0 | CPULoadMonitor
	  --------------------------------------------------

	  After this commit , output is : 
          Started CPU monitor with interval 5 sec.
	  Non-Zero CPU utilization trend (updated every 2s)
          --------------------------------------------------
	  PID | Pri |  CPU0 |  CPU1 |  2s  Avg|
	  --------------------------------------------------
 	    0 |   0 |  99.9 |   0.0 |  42.8 | CPU0 IDLE
 	    1 |   0 |   0.0 | 100.0 |  57.2 | CPU1 IDLE
 	   20 | 100 |   1.4 |   0.0 |   0.6 | CPULoadMonitor
	  --------------------------------------------------
```

edwakuwaku commented:
What does "2s" refer to?

ritesh55555 commented:
> What does "2s" refer to?

2s refer to 2 sec , so the cpu load is calculated for last 2 sec (it represents) . The 2s refers to total cpu utilisation(considering both cpu0 and cpu1). I think we should change the name but it was there before so i did not change it

sunghan-chang commented:
> > What does "2s" refer to?
> 
> 2s refer to 2 sec , so the cpu load is calculated for last 2 sec (it represents) . The 2s refers to total cpu utilisation(considering both cpu0 and cpu1). I think we should change the name but it was there before so i did not change it

1. Oh?? It just calculates cpu load for last 2 sec?? Some thread works at first 3 sec, then it shows 0?? It looks much strange. Could you check it again?

2. When you modify `2s`, could you check `TASH>> cpuload Started CPU monitor with interval 5`? interval 5 means? 5 sec? 5 min?

ritesh55555 commented:
> > > What does "2s" refer to?
> > 
> > 
> > 2s refer to 2 sec , so the cpu load is calculated for last 2 sec (it represents) . The 2s refers to total cpu utilisation(considering both cpu0 and cpu1). I think we should change the name but it was there before so i did not change it
> 
> 1. Oh?? It just calculates cpu load for last 2 sec?? Some thread works at first 3 sec, then it shows 0?? It looks much strange. Could you check it again?
> 2. When you modify `2s`, could you check `TASH>> cpuload Started CPU monitor with interval 5`? interval 5 means? 5 sec? 5 min?

1. Yes , it always calculates the cpu load for last 2 seconds . (But this 2 second value can be modified from config). After 2 second, kernel decreases each load of threads by half and total load by half and the process continues .
```
if (g_cpuload_total[cpuload_idx] > (g_cpuload_timeconstant[cpuload_idx] * CPULOAD_TICKSPERSEC)) {
			uint32_t total = 0;

			/* Divide the tick count for every task by two and recalculate the
			 * total.
			 */
			for (i = 0; i < CONFIG_MAX_TASKS; i++) {
				g_pidhash[i].ticks[cpuload_idx] >>= 1;
				total += g_pidhash[i].ticks[cpuload_idx];
			}
			/* Save the new total. */

			g_cpuload_total[cpuload_idx] = total;
		}
```
So yes if a thread works at first 3 sec  and then we call cpuload command after 5 sec , then it will show 0 load for the thread.
There is one more config of multi cpuload ,where the output shows three columns, one for last 2sed, one for last 5 sec and one for last 10 sec . Again this values can be changed from configs.


2. interval 5 means it will show the result logs after every 5 second but output will contain data of last 2 second

sunghan-chang commented:
> > > > What does "2s" refer to?
> > > 
> > > 
> > > 2s refer to 2 sec , so the cpu load is calculated for last 2 sec (it represents) . The 2s refers to total cpu utilisation(considering both cpu0 and cpu1). I think we should change the name but it was there before so i did not change it
> > 
> > 
> > 
> > 1. Oh?? It just calculates cpu load for last 2 sec?? Some thread works at first 3 sec, then it shows 0?? It looks much strange. Could you check it again?
> > 2. When you modify `2s`, could you check `TASH>> cpuload Started CPU monitor with interval 5`? interval 5 means? 5 sec? 5 min?
> 
> hi @sunghan-chang
> 
> 1. Yes , it always calculates the cpu load for last 2 seconds . (But this 2 second value can be modified from config CONFIG_SCHED_CPULOAD_TIMECONSTANT ). After 2 second, kernel decreases each load of threads by half and total load by half and the process continues .
> 
> ```
> if (g_cpuload_total[cpuload_idx] > (g_cpuload_timeconstant[cpuload_idx] * CPULOAD_TICKSPERSEC)) {
> 			uint32_t total = 0;
> 
> 			/* Divide the tick count for every task by two and recalculate the
> 			 * total.
> 			 */
> 			for (i = 0; i < CONFIG_MAX_TASKS; i++) {
> 				g_pidhash[i].ticks[cpuload_idx] >>= 1;
> 				total += g_pidhash[i].ticks[cpuload_idx];
> 			}
> 			/* Save the new total. */
> 
> 			g_cpuload_total[cpuload_idx] = total;
> 		}
> ```
> 
> So yes if a thread works at first 3 sec and then we call cpuload command after 5 sec , then it will show 0 load for the thread. There is one more config of multi cpuload ,where the output shows three columns, one for last 2sed, one for last 5 sec and one for last 10 sec . Again this values can be changed from configs.
> 
> 2. interval 5 means it will show the result logs after every 5 second but output will contain data of last 2 second
>    here 5sec = CONFIG_CPULOADMONITOR_INTERVAL
>    2sec = CONFIG_SCHED_CPULOAD_TIMECONSTANT

1. Yes, every few seconds, previous calculated load becomes half. But it does not be zero. I mean, cpu load of last 2 second applies 100%, but it does not mean it just calculated only last 2 secs.
2. Yes, I know. What I mean is the sentence should say the unit of `5`. The sentence is for user who does not know cpu load code itself. If you say just `5`, user can't know the unit of it.

ritesh55555 commented:
> > > > > What does "2s" refer to?
> > > > 
> > > > 
> > > > 2s refer to 2 sec , so the cpu load is calculated for last 2 sec (it represents) . The 2s refers to total cpu utilisation(considering both cpu0 and cpu1). I think we should change the name but it was there before so i did not change it
> > > 
> > > 
> > > 
> > > 1. Oh?? It just calculates cpu load for last 2 sec?? Some thread works at first 3 sec, then it shows 0?? It looks much strange. Could you check it again?
> > > 2. When you modify `2s`, could you check `TASH>> cpuload Started CPU monitor with interval 5`? interval 5 means? 5 sec? 5 min?
> > 
> > 
> > hi @sunghan-chang
> > 
> > 1. Yes , it always calculates the cpu load for last 2 seconds . (But this 2 second value can be modified from config CONFIG_SCHED_CPULOAD_TIMECONSTANT ). After 2 second, kernel decreases each load of threads by half and total load by half and the process continues .
> > 
> > ```
> > if (g_cpuload_total[cpuload_idx] > (g_cpuload_timeconstant[cpuload_idx] * CPULOAD_TICKSPERSEC)) {
> > 			uint32_t total = 0;
> > 
> > 			/* Divide the tick count for every task by two and recalculate the
> > 			 * total.
> > 			 */
> > 			for (i = 0; i < CONFIG_MAX_TASKS; i++) {
> > 				g_pidhash[i].ticks[cpuload_idx] >>= 1;
> > 				total += g_pidhash[i].ticks[cpuload_idx];
> > 			}
> > 			/* Save the new total. */
> > 
> > 			g_cpuload_total[cpuload_idx] = total;
> > 		}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So yes if a thread works at first 3 sec and then we call cpuload command after 5 sec , then it will show 0 load for the thread. There is one more config of multi cpuload ,where the output shows three columns, one for last 2sed, one for last 5 sec and one for last 10 sec . Again this values can be changed from configs.
> > 
> > 2. interval 5 means it will show the result logs after every 5 second but output will contain data of last 2 second
> >    here 5sec = CONFIG_CPULOADMONITOR_INTERVAL
> >    2sec = CONFIG_SCHED_CPULOAD_TIMECONSTANT
> 
> 1. Yes, every few seconds, previous calculated load becomes half. But it does not become zero. I mean, cpu load of last 2 second applies 100%, but it does not mean it just calculated only last 2 secs.
> 2. Yes, I know. What I mean is the sentence should say the unit of `5`. The sentence is for user who does not know cpu load code itself. If you say just `5`, user can't know the unit of it.

1 . Right. Sorry for the confusion . it does not exactly give the last 2 seconds data .  But after every 2 seconds , it reduces each threads ticks count by half (even if the thread was started way before 2 second) and accordingly calculates the cpu load . 
Is there a requirement of different way of calculating the load value?

2. updated!

sunghan-chang commented:
> > > > > > What does "2s" refer to?
> > > > > 
> > > > > 
> > > > > 2s refer to 2 sec , so the cpu load is calculated for last 2 sec (it represents) . The 2s refers to total cpu utilisation(considering both cpu0 and cpu1). I think we should change the name but it was there before so i did not change it
> > > > 
> > > > 
> > > > 
> > > > 1. Oh?? It just calculates cpu load for last 2 sec?? Some thread works at first 3 sec, then it shows 0?? It looks much strange. Could you check it again?
> > > > 2. When you modify `2s`, could you check `TASH>> cpuload Started CPU monitor with interval 5`? interval 5 means? 5 sec? 5 min?
> > > 
> > > 
> > > hi @sunghan-chang
> > > 
> > > 1. Yes , it always calculates the cpu load for last 2 seconds . (But this 2 second value can be modified from config CONFIG_SCHED_CPULOAD_TIMECONSTANT ). After 2 second, kernel decreases each load of threads by half and total load by half and the process continues .
> > > 
> > > ```
> > > if (g_cpuload_total[cpuload_idx] > (g_cpuload_timeconstant[cpuload_idx] * CPULOAD_TICKSPERSEC)) {
> > > 			uint32_t total = 0;
> > > 
> > > 			/* Divide the tick count for every task by two and recalculate the
> > > 			 * total.
> > > 			 */
> > > 			for (i = 0; i < CONFIG_MAX_TASKS; i++) {
> > > 				g_pidhash[i].ticks[cpuload_idx] >>= 1;
> > > 				total += g_pidhash[i].ticks[cpuload_idx];
> > > 			}
> > > 			/* Save the new total. */
> > > 
> > > 			g_cpuload_total[cpuload_idx] = total;
> > > 		}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > So yes if a thread works at first 3 sec and then we call cpuload command after 5 sec , then it will show 0 load for the thread. There is one more config of multi cpuload ,where the output shows three columns, one for last 2sed, one for last 5 sec and one for last 10 sec . Again this values can be changed from configs.
> > > 
> > > 2. interval 5 means it will show the result logs after every 5 second but output will contain data of last 2 second
> > >    here 5sec = CONFIG_CPULOADMONITOR_INTERVAL
> > >    2sec = CONFIG_SCHED_CPULOAD_TIMECONSTANT
> > 
> > 
> > 
> > 1. Yes, every few seconds, previous calculated load becomes half. But it does not become zero. I mean, cpu load of last 2 second applies 100%, but it does not mean it just calculated only last 2 secs.
> > 2. Yes, I know. What I mean is the sentence should say the unit of `5`. The sentence is for user who does not know cpu load code itself. If you say just `5`, user can't know the unit of it.
> 
> 1 . Right. Sorry for the confusion . it does not exactly give the last 2 seconds data . But after every 2 seconds , it reduces each threads ticks count by half (even if the thread was started way before 2 second) and accordingly calculates the cpu load . Is there a requirement of different way of calculating the load value?
> 
> 2. updated!

1. No. It's ok. I just want to get confirmation. But need user to know that. Could you write the way of calculation in the README of cpuload? It's in apps/system/utils. And as said above, need to change `2s` title in the chart.

kishore-sn commented on apps/system/utils/utils_cpuload.c:
```diff
@@ -136,14 +161,35 @@ static void cpuload_print_pid_value(char *buf, void *arg)
 	} else {
 #ifdef CONFIG_SCHED_MULTI_CPULOAD
-		printf(" %5s | %5s | %5s |", stat_info[PROC_STAT_CPULOAD_SHORT], stat_info[PROC_STAT_CPULOAD_MID], stat_info[PROC_STAT_CPULOAD_LONG]);
+#ifdef CONFIG_SMP
```
There is no need to maintain a separate code for SMP case as NCPU will always be set to 1 if smp is not present.

edwakuwaku commented on apps/system/utils/utils_cpuload.c:
```diff
@@ -123,6 +141,13 @@ static void cpuload_print_pid_value(char *buf, void *arg)
 		stat_info[i] = strtok_r(stat_info[i], " ", &stat_info[i + 1]);
 	}
 
+#ifdef CONFIG_SCHED_MULTI_CPULOAD
```
```
#ifdef CONFIG_SCHED_MULTI_CPULOAD
	if (!(has_cpuload(stat_info[PROC_STAT_CPULOAD_SHORT]) || has_cpuload(stat_info[PROC_STAT_CPULOAD_MID]) || has_cpuload(stat_info[PROC_STAT_CPULOAD_LONG]))) {
#else 
	if (!has_cpuload(stat_info[PROC_STAT_CPULOAD])) {
#endif
		return;
        }
```

kishore-sn commented on os/kernel/task/task_setup.c:
```diff
@@ -173,7 +173,9 @@ static int task_assignpid(FAR struct tcb_s *tcb)
 			int cpuload_idx;
 			for (cpuload_idx = 0; cpuload_idx < SCHED_NCPULOAD; cpuload_idx++) {
-				g_pidhash[hash_ndx].ticks[cpuload_idx] = 0;
+				for (int i = 0; i < CONFIG_SMP_NCPUS; i++) {
```
Instead of using variable name as i, we can name it as cpu to make it more meaningful.
You can change accordingly at all places in your pr.

kishore-sn commented on os/kernel/sched/sched_cpuload.c:
```diff
@@ -286,21 +291,22 @@ void weak_function sched_process_cpuload(void)
 		/* Increment tick count.  If the accumulated tick value exceed a time
 		 * constant, then shift the accumulators.
 		 */
+		for (i = 0; i < CONFIG_SMP_NCPUS; i++) {
```
We can combine this for loop with the above one at line 287.

kishore-sn commented on os/fs/procfs/fs_procfsproc.c:
```diff
@@ -363,6 +363,7 @@ static ssize_t proc_entry_stat(FAR struct proc_file_s *procfile, FAR struct tcb_
 	size_t peak_heap;
 	size_t peak_stack;
 	pid_t ppid;
+
```
Revert unnecessary empty line.

kishore-sn commented on os/fs/procfs/fs_procfsproc.c:
```diff
@@ -407,27 +410,49 @@ static ssize_t proc_entry_stat(FAR struct proc_file_s *procfile, FAR struct tcb_
+
+		}
+
+		if (total_cpuload > 0) {
```
`total_cpuload` will be the total load of all the cpu in the system. However, its value needs to be between 0 - 100%. So, we need to take average of all the cpu loads here. 
`total_cpuload = total_cpuload / CONFIG_SMP_NCPUS;`

Please change accordingly at all required points.

kishore-sn commented on os/fs/procfs/fs_procfsproc.c:
```diff
@@ -666,22 +693,26 @@ static ssize_t proc_entry_loadavg(FAR struct proc_file_s *procfile, FAR struct t
 		(void)clock_cpuload(procfile->pid, cpuload_idx, &cpuload);
 
+		for (int i = 0; i < CONFIG_SMP_NCPUS; i++) {
+			total_cpuload += cpuload.total[i];
```
Same as previous comment. Need to divide by ncpu after the for loop.

kishore-sn commented on apps/system/utils/utils_cpuload.c:
```diff
@@ -62,6 +62,14 @@
+	CONFIG_SCHED_CPULOAD_TIMECONSTANT_SHORT,
+	CONFIG_SCHED_CPULOAD_TIMECONSTANT_MID,
+	CONFIG_SCHED_CPULOAD_TIMECONSTANT_LONG
+};
```
Does this need to be a global variable? Can we move it into the function where its being used?

kishore-sn commented on apps/system/utils/README.md:
```diff
@@ -126,7 +126,11 @@ Kernel Features -> Files and I/O -> Maximum number of file descriptors per task
 ## cpuload
-This command shows cpuload information per thread periodically. This has arguments which configure the printing information and cpuload daemon.
+This command shows cpuload information per thread periodically. 
+For SMP case, it shows cpuload of thread in each core.
```
`For SMP case` ===> `If there are more than one cpu cores in the system`

kishore-sn commented on apps/system/utils/README.md:
```diff
@@ -140,27 +144,19 @@ Options:
+--------------------------------------------------
+Started CPU monitor with interval 5 sec.
+TASH>>Non-Zero CPU utilization (2s)
+PID | Pri |  CPU0 |  CPU1 | 2s Avg|
```
We dont have to mention 2s Avg here. All the values (each cpu value) are averaged over 2s.

kishore-sn commented on apps/system/utils/README.md:
```diff
@@ -183,6 +179,16 @@ PID | Pri |  Snap ticks  |
+### How is the cpuload calculated
+We define a timeconstant (CONFIG_SCHED_CPULOAD_TIMECONSTANT), to keep a limit on total cpuload.
+We store "total ticks" and "ticks respective to a thread for specific core" after every systick interrupt.
+thread_n load on cpu_i = (thread_n load in cpu_i) / (total load in cpu_i)
```
`thread_n load on cpu_i = (thread_n load in cpu_i) / (total load in cpu_i)` 
Change to 
`thread_n load on cpu_i = (thread_n ticks in cpu_i) / (total ticks in cpu_i)`

kishore-sn commented on apps/system/utils/README.md:
```diff
@@ -183,6 +179,16 @@ PID | Pri |  Snap ticks  |
+We define a timeconstant (CONFIG_SCHED_CPULOAD_TIMECONSTANT), to keep a limit on total cpuload.
+We store "total ticks" and "ticks respective to a thread for specific core" after every systick interrupt.
+thread_n load on cpu_i = (thread_n load in cpu_i) / (total load in cpu_i)
+average thread_n load = (thread_n load in all cpu) / (total load in all cpu)
```
`average thread_n load = (thread_n load in all cpu) / (total load in all cpu)`
Change to
`average thread_n load = (Sum of thread_n load in each cpu) / (Number of cpus)`

kishore-sn commented:
@ritesh55555 
After addressing the above comments, pls check the following:

Start cpuload on device
Start smp test app
While spm app is running, the cpuload is printed. 
Paste that output here.

ritesh55555 commented:
> @ritesh55555 After addressing the above comments, pls check the following:
> 
> Start cpuload on device Start smp test app While spm app is running, the cpuload is printed. Paste that output here.

Thread[4]: Done
Thread[8]: Now running on CPU0
Thread[7]: Now running on CPU0
Thread[6]: Now running on CPU0
Thread[1]: Now running on CPU0
Thread[2]: Now running on CPU1
Thread[3]: Done
Thread[6]: Done

--------------------------------------------------
Non-Zero CPU utilization (2s)
PID | Pri |  CPU0 |  CPU1 |  Avg  |
--------------------------------------------------
  0 |   0 |   2.5 |   0.0 |   1.2 | CPU0 IDLE
  1 |   0 |   0.0 |   0.4 |   0.2 | CPU1 IDLE
 31 | 100 |   0.2 |   0.0 |   0.1 | CPULoadMonitor
 33 | 100 |  25.7 |   7.6 |  16.2 | <pthread>
 34 | 100 |  11.3 |  15.9 |  13.7 | <pthread>
 35 | 100 |   8.2 |  16.6 |  12.6 | <pthread>
 37 | 100 |  18.7 |  14.4 |  16.5 | <pthread>
 38 | 100 |  11.1 |  13.0 |  12.1 | <pthread>
 39 | 100 |  14.0 |  13.6 |  13.8 | <pthread>
 40 | 100 |   8.1 |  18.6 |  13.6 | <pthread>
--------------------------------------------------

kishore-sn commented on os/fs/procfs/fs_procfsproc.c:
```diff
@@ -666,22 +692,26 @@ static ssize_t proc_entry_loadavg(FAR struct proc_file_s *procfile, FAR struct t
 			uint32_t tmp;
-
-			tmp = (1000 * cpuload.active) / cpuload.total;
+			tmp = 1000 - (1000 * total_active) / total_cpuload;
```
I think you need to remove `1000 -` in the above line.

kishore-sn commented on os/fs/procfs/fs_procfsproc.c:
```diff
@@ -666,22 +692,26 @@ static ssize_t proc_entry_loadavg(FAR struct proc_file_s *procfile, FAR struct t
 		}
 
-		linesize = snprintf(procfile->line, STATUS_LINELEN, "%3d.%01d%%   ", intpart, fracpart);
+		linesize = snprintf(procfile->line, STATUS_LINELEN, "%3d.%01d%%   ", intpart, fracpart);		
```
Remove space at end of line.

kishore-sn commented on os/fs/procfs/fs_procfscpuload.c:
```diff
@@ -253,15 +255,19 @@ static ssize_t cpuload_read(FAR struct file *filep, FAR char *buffer, size_t buf
 		for (cpuload_idx = 0; cpuload_idx < SCHED_NCPULOAD; cpuload_idx++) {
 
 			DEBUGVERIFY(clock_cpuload(0, cpuload_idx, &cpuload));
+			for (int i = 0; i < CONFIG_SMP_NCPUS; i++) {
```
i -> cpu

kishore-sn commented on apps/system/utils/utils_cpuload.c:
```diff
@@ -109,6 +109,16 @@ static void cpuload_stop(void)
 
+static bool has_cpuload(char *load_info)
+{
+	int len = strlen(load_info);
```
Pls add  a comment here about why you are comparing with "-0.0". If someone changes the format of cpuload, then this will get affected.

sunghan-chang commented:
`

> > @ritesh55555 After addressing the above comments, pls check the following:
> > Start cpuload on device Start smp test app While spm app is running, the cpuload is printed. Paste that output here.
> 
> ```
> Thread[4]: Done
> Thread[8]: Now running on CPU0
> Thread[7]: Now running on CPU0
> Thread[6]: Now running on CPU0
> Thread[1]: Now running on CPU0
> Thread[2]: Now running on CPU1
> Thread[3]: Done
> Thread[6]: Done
> 
> --------------------------------------------------
> Non-Zero CPU utilization (2s)
> PID | Pri |  CPU0 |  CPU1 |  Avg  |
> --------------------------------------------------
>   0 |   0 |   2.5 |   0.0 |   1.2 | CPU0 IDLE
>   1 |   0 |   0.0 |   0.4 |   0.2 | CPU1 IDLE
>  31 | 100 |   0.2 |   0.0 |   0.1 | CPULoadMonitor
>  33 | 100 |  25.7 |   7.6 |  16.2 | <pthread>
>  34 | 100 |  11.3 |  15.9 |  13.7 | <pthread>
>  35 | 100 |   8.2 |  16.6 |  12.6 | <pthread>
>  37 | 100 |  18.7 |  14.4 |  16.5 | <pthread>
>  38 | 100 |  11.1 |  13.0 |  12.1 | <pthread>
>  39 | 100 |  14.0 |  13.6 |  13.8 | <pthread>
>  40 | 100 |   8.1 |  18.6 |  13.6 | <pthread>
> --------------------------------------------------
> ```

What does it mean `(2s)`? Same as before? Result of Last 2s?

ritesh55555 commented:
> `
> 
> > > @ritesh55555 After addressing the above comments, pls check the following:
> > > Start cpuload on device Start smp test app While spm app is running, the cpuload is printed. Paste that output here.
> > 
> > 
> > ```
> > Thread[4]: Done
> > Thread[8]: Now running on CPU0
> > Thread[7]: Now running on CPU0
> > Thread[6]: Now running on CPU0
> > Thread[1]: Now running on CPU0
> > Thread[2]: Now running on CPU1
> > Thread[3]: Done
> > Thread[6]: Done
> > 
> > --------------------------------------------------
> > Non-Zero CPU utilization (2s)
> > PID | Pri |  CPU0 |  CPU1 |  Avg  |
> > --------------------------------------------------
> >   0 |   0 |   2.5 |   0.0 |   1.2 | CPU0 IDLE
> >   1 |   0 |   0.0 |   0.4 |   0.2 | CPU1 IDLE
> >  31 | 100 |   0.2 |   0.0 |   0.1 | CPULoadMonitor
> >  33 | 100 |  25.7 |   7.6 |  16.2 | <pthread>
> >  34 | 100 |  11.3 |  15.9 |  13.7 | <pthread>
> >  35 | 100 |   8.2 |  16.6 |  12.6 | <pthread>
> >  37 | 100 |  18.7 |  14.4 |  16.5 | <pthread>
> >  38 | 100 |  11.1 |  13.0 |  12.1 | <pthread>
> >  39 | 100 |  14.0 |  13.6 |  13.8 | <pthread>
> >  40 | 100 |   8.1 |  18.6 |  13.6 | <pthread>
> > --------------------------------------------------
> > ```
> 
> What does it mean `(2s)`? Same as before? Result of Last 2s?

Yes

sunghan-chang commented:
@ritesh55555 Current sentence is not understandable. That sentence is for user, not for you. So you should change it to know easily for who does not use cpuload code.

ritesh55555 commented:
> Non-Zero CPU utilization trend (updated every 2s)

i have updated it to  Non-Zero CPU utilization trend (updated every 2s)" .

kishore-sn commented on apps/system/utils/README.md:
```diff
@@ -140,27 +144,19 @@ Options:
+
+--------------------------------------------------
+Started CPU monitor with interval 5 sec.
+TASH>>Non-Zero CPU utilization (2s)
```
Can you pls check what is 2s and what is 5 sec here? 
As i understand. we will display the cpu load info every 5 secs. 
At the end of every 2 secs the accumulated ticks value is divided by 2. This is an internal aspect of the cpuload utility to prevent overflow of data. The user does not need to know this information. So, we can remove it.

ritesh55555 replied:
i have updated the line with "Non-Zero CPU utilization trend (updated every 5 sec)" because we will see the estimated average value after every 5 sec (or cpuload interval for showing data)

kishore-sn commented on apps/system/utils/utils_cpuload.c:
```diff
@@ -136,14 +161,27 @@ static void cpuload_print_pid_value(char *buf, void *arg)
 	} else {
 #ifdef CONFIG_SCHED_MULTI_CPULOAD
-		printf(" %5s | %5s | %5s |", stat_info[PROC_STAT_CPULOAD_SHORT], stat_info[PROC_STAT_CPULOAD_MID], stat_info[PROC_STAT_CPULOAD_LONG]);
+	for (i = PROC_STAT_CPULOAD_SHORT; i <= PROC_STAT_CPULOAD_LONG; i++) {
```
The indentation seems to be wrong.

sunghan-chang commented on apps/system/utils/utils_cpuload.c:
```diff
@@ -214,33 +205,25 @@ static void cpuload_print_normal(void)
+	printf("Non-Zero CPU utilization trend (updated every %ds)\n", cpuload_interval);
 #else
-	printf("Non-Zero CPU utilization (%ds)\n", CONFIG_SCHED_CPULOAD_TIMECONSTANT);
+	printf("Non-Zero CPU utilization trend (updated every %ds)\n", cpuload_interval);
```
The `ifdef` and the `else` cases are perfectly the same. Let's keep one line for this.

sunghan-chang commented on apps/system/utils/utils_cpuload.c:
```diff
@@ -159,19 +197,35 @@ static int cpuload_read_proc(FAR struct dirent *entryp, FAR void *arg)
+
 #ifdef CONFIG_SCHED_MULTI_CPULOAD
-	printf("%5ds | %4ds | %4ds |", CONFIG_SCHED_CPULOAD_TIMECONSTANT_SHORT, CONFIG_SCHED_CPULOAD_TIMECONSTANT_MID, CONFIG_SCHED_CPULOAD_TIMECONSTANT_LONG);
+	for (int i = PROC_STAT_CPULOAD_SHORT; i <= PROC_STAT_CPULOAD_LONG; i++) {
+		for (int j = 0; j < CONFIG_SMP_NCPUS; j++) {
+			printf("  CPU%d |", j);
+		}
+		printf("  Avg  |");
+	}
+
 #else
-	printf("%5ds |", CONFIG_SCHED_CPULOAD_TIMECONSTANT);
+	for (int i = 0; i < CONFIG_SMP_NCPUS; i++) {
+		printf("  CPU%d |", i);
+	}
+	printf("  Avg  |");
 #endif
```
It looks we can simplify the code like below
```
#ifdef CONFIG_.....
for (int j = PROC....) {
#endif
    for (int i = 0; ....) {
        printf("    cpu%d....);
    }
    printf("    avg....);
#ifdef CONFIG_.....
}
#endif
```

kishore-sn commented on apps/system/utils/utils_cpuload.c:
```diff
@@ -159,19 +197,27 @@ static int cpuload_read_proc(FAR struct dirent *entryp, FAR void *arg)
-	printf("%5ds | %4ds | %4ds |", CONFIG_SCHED_CPULOAD_TIMECONSTANT_SHORT, CONFIG_SCHED_CPULOAD_TIMECONSTANT_MID, CONFIG_SCHED_CPULOAD_TIMECONSTANT_LONG);
-#else
-	printf("%5ds |", CONFIG_SCHED_CPULOAD_TIMECONSTANT);
+	for (int i = PROC_STAT_CPULOAD_SHORT; i <= PROC_STAT_CPULOAD_LONG; i++) {
```
This is not the right way of printing. When multiload is enabled in single core scenario, then we print the load for 3 cases, short, mid and long time duration. When we have multi-core, then for each duration, we need to print load of each cpu and avg load. Please recheck this - the table heading as well as the data being printed. Once this is done, pls paste a sample output of multiload in the commit.

ritesh55555 replied:
This is for single core case :
```
--------------------------------------------------
Non-Zero CPU utilization trend (updated every 5s)
PID | Pri |  CPU0 |  Avg  |  CPU0 |  Avg  |  CPU0 |  Avg  |
--------------------------------------------------
  0 |   0 |  99.9 |  99.9 |  99.8 |  99.8 |  99.8 |  99.8 | Idle Task
 11 | 203 |   0.0 |   0.0 |   0.1 |   0.0 |   0.1 |   0.1 | binary_manager
 14 | 180 |   0.0 |   0.0 |   0.1 |   0.1 |   0.1 |   0.1 | app1
 19 | 100 |   0.2 |   0.2 |   0.1 |   0.1 |   0.1 |   0.1 | CPULoadMonitor
--------------------------------------------------
```

This is for multi core case:
```
--------------------------------------------------
Non-Zero CPU utilization trend (updated every 5s)
PID | Pri |  CPU0 |  CPU1 |  Avg  |  CPU0 |  CPU1 |  Avg  |  CPU0 |  CPU1 |  Avg  |
--------------------------------------------------
  0 |   0 |  99.9 |   0.0 |  49.7 | 100.0 |   0.0 |  49.3 |  99.9 |   0.0 |  49.4 | CPU0 IDLE
  1 |   0 |   0.0 |  99.9 |  50.3 |   0.0 |  99.8 |  50.6 |   0.0 |  99.8 |  50.5 | CPU1 IDLE
 29 | 100 |   0.2 |   0.1 |   0.1 |   0.1 |   0.1 |   0.1 |   0.1 |   0.1 |   0.1 | CPULoadMonitor
--------------------------------------------------
```

kishore-sn replied:
In my opinion, multiload scenario will not be used often and we dont have to give details per-cpu information in multiload case. So, for multiload case, let us keep the output simple as follows:
```
--------------------------------------------------
Non-Zero CPU utilization trend (updated every 5s)
PID | Pri |  (%d sec) Avg  |   (%d sec)  Avg  |   (%d sec)   Avg  |
--------------------------------------------------
```

ritesh55555 replied:
These are the updated output . By adding (%d sec) Avg , there was slightly more code added and the output look very spacious , so i updated like this . Is it okay?
```
--------------------------------------------------
Non-Zero CPU utilization trend (updated every 5s)
Multi cpuload mode is enabled with time constants 2s, 5s, 10s (used to average out cpuload after every time constants)
PID | Pri |  Avg1 |  Avg2 |  Avg3 |
--------------------------------------------------
  0 |   0 |  99.8 |  99.7 |  99.7 | Idle Task
 11 | 203 |   0.1 |   0.1 |   0.1 | binary_manager
 14 | 180 |   0.1 |   0.1 |   0.1 | app1
 19 | 100 |   0.2 |   0.1 |   0.1 | CPULoadMonitor
--------------------------------------------------


--------------------------------------------------
Non-Zero CPU utilization trend (updated every 5s)
Multi cpuload mode is enabled with time constants 2s, 5s, 10s (used to average out cpuload after every time constants)
PID | Pri |  CPU0 |  CPU1 |  Avg1 |  CPU0 |  CPU1 |  Avg2 |  CPU0 |  CPU1 |  Avg3 |
--------------------------------------------------
  0 |   0 |  99.8 |   0.0 |  47.9 |  99.9 |   0.0 |  49.0 |  99.9 |   0.0 |  49.3 | CPU0 IDLE
  1 |   0 |   0.0 | 100.0 |  52.0 |   0.0 | 100.0 |  51.0 |   0.0 | 100.0 |  50.6 | CPU1 IDLE
 20 | 100 |   0.4 |   0.0 |   0.2 |   0.1 |   0.0 |   0.1 |   0.1 |   0.0 |   0.1 | CPULoadMonitor
--------------------------------------------------


--------------------------------------------------
Non-Zero CPU utilization trend (updated every 5s)
PID | Pri |  CPU0 |  CPU1 |  Avg  |
--------------------------------------------------
  0 |   0 |  99.9 |   0.0 |  48.6 | CPU0 IDLE
  1 |   0 |   0.0 | 100.0 |  51.3 | CPU1 IDLE
 20 | 100 |   0.1 |   0.0 |   0.1 | CPULoadMonitor
--------------------------------------------------

--------------------------------------------------
Non-Zero CPU utilization trend (updated every 5s)
PID | Pri |  Avg  |
--------------------------------------------------
  0 |   0 |  99.9 | Idle Task
 19 | 100 |   0.1 | CPULoadMonitor
--------------------------------------------------
```

kishore-sn replied:
Looks fine.
